libcontainer/factory_linux.go

```diff
@@ -378,6 +379,31 @@ func (l *LinuxFactory) StartInitialization() (err error) {
 		return fmt.Errorf("unable to convert _LIBCONTAINER_LOGPIPE: %w", err)
 	}
 
+	// Get mount files (O_PATH).
+	// Keep mountFds as nil if there are no fds to mount. Note we assume (and verify!) this in other parts of the code.
+	var mountFds []*int
+	fdsStr := os.Getenv("_LIBCONTAINER_MOUNT_FDS")
+	if fdsStr != "" {
+		// This string was created so we can parse it just fine with strings.Split() and it has the proper len.
+		fds := strings.Split(fdsStr, ";")
+
+		// Indexes in the mountFds slice are used to assign/pair the fd (if any) to a mount in the container. Therefore,
+		// the mountFds slice MUST be of size len(fds), that was constructed with the same size as the container mounts.
+		mountFds = make([]*int, len(fds))
+		for i, fd := range fds {
+			if fd == "" {
+				continue
+			}
+
```

Comment:
Well, the "we can't be sure to initialize all of them to -1" is something you said ;). My point is that not having the same number of entries the mountFds slice as the container mounts makes it difficult to pair and assign one fd to a mount :)

What do you think is saner about having -1 in the data structure? Honestly, I'm not sure I see what you think is saner.

IMHO, it doesn't look saner. Later, when we do use the fds, we or others in the future might check "is -1?" and if not, use them. While unlikely, if someone stores some other invalid fd (like anything <0, -2, etc.) if we are checking for `!= -1`that will cause problems, etc. IMHO, it seems way simpler to have only one possible representation to say "don't use an fd" (i.e. use nil) and if you do add a pointer, it must be a valid fd. Those semantics seems clearer to me and less error prone.

What am I missing?

Reply:
To reduce the RTT I just switched this to -1 as you want. I'm pushing this as a commit on top, just so you can easily see the diff. Please feel free to squash it or whatever if you want to merge this ASAP.

Some things I noted when doing the change, that makes me think it doesn't smell quite good:
* I just switched from []*int to []int in the functions parsing the _LIBCONTAINER_MOUNT_FDS vars. For `mountConfig` that has a `fd` field, I kepts it as `*int`. Changing that to `int` would mean that it is 0 initialized and asking for bugs to happen. It doesn't seem okay to use `int` that the zero value is problematic for our use and therefore forcing all users to specify -1 when they create the struct. That seems like an anti-pattern in go, so kept *int that works just fine with the 0 value and is not so error prone.
* Being a `[]int` is not as clearer as `*int` that you might not be able to use the int as an fd. For example [here](https://github.com/opencontainers/runc/pull/2576/commits/16d6957177410530895fae9276534f3d1f8399a3#diff-cdcc9c0684f1dd890d73f0a782b7d878b51996f83262ea72a48c1ebdc7a7d21eR95) or [here](https://github.com/opencontainers/runc/pull/2576/commits/16d6957177410530895fae9276534f3d1f8399a3#diff-e0932af06cb46065b2ef5fc89bc2bd9b880c9578e01b89ef4ecaeffb78a8bef0R82). Whereas, if it a pointer it is quite clear that you need to check for nil and otherwise you can use it.

Let me know what you think @kolyshkin

Reply:
So, we need a flag to say "don't use this fd". Since the fd is a non-negative number, we can use the sign bit of an integer for this flag.

OTOH the pointer is an extra level of indirection, and involves an extra dynamic allocation. go runtime has to allocate an integer, assign the pointer to the location of that integer, and dereference the pointer each time we want to access it. It also means it has to store this "we have allocated an object which needs to be freed" in its garbage collector data somewhere, and actually free each individual integer once it's out of scope.

So, using a `[]*int` instead of `[]int` means instead of using a single bit (which is unused otherwise) for a flag, we use 64 bits, more memory allocations, dereferencing, and incur GC overhead.

Sure, if we can go back in time 30 years I'd say "let's start fds from 1 not 0" and the whole problem is solved. Alas, we can not.